File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.py
def solution(places):
    answer = []
    
    for i in range(5):
        room = places[i]
        covid = False
        
        for a in range(5):
            if covid:
                break
            for b in range(5):
                if room[a][b] == 'P':
                    if 0 <= a < 5 and 0 <= b - 1 < 5 and room[a][b-1] == 'P':
                        answer.append(0)
                        covid = True
                        break
                    if 0 <= a - 1 < 5 and 0 <= b < 5 and room[a-1][b] == 'P':
                        answer.append(0)
                        covid = True
                        break
                    if 0 <= a < 5 and 0 <= b + 1 < 5 and room[a][b+1] == 'P':
                        answer.append(0)
                        covid = True
                        break
                    if 0 <= a + 1 < 5 and 0 <= b < 5 and room[a+1][b] == 'P':
                        answer.append(0)
                        covid = True
                        break
                    if 0 <= a < 5 and 0 <= b - 2 < 5 and room[a][b - 2] == 'P':
                        if room[a][b-1] == 'O':
                            answer.append(0)
                            covid = True
                            break
                    if 0 <= a - 2 < 5 and 0 <= b < 5 and room[a-2][b] == 'P':
                        if room[a-1][b] == 'O':
                            answer.append(0)
                            covid = True
                            break
                    if 0 <= a < 5 and 0 <= b + 2 < 5 and room[a][b+2] == 'P':
                        if room[a][b+1] == 'O':
                            answer.append(0)
                            covid = True
                            break
                    if 0 <= a + 2 < 5 and 0 <= b < 5 and room[a+2][b] == 'P':
                        if room[a+1][b] == 'O':
                            answer.append(0)
                            covid = True
                            break
                    if 0 <= a - 1 < 5 and 0 <= b - 1 < 5 and room[a-1][b-1] == 'P':
                        if room[a-1][b] == 'O' or room[a][b-1] == 'O':
                            answer.append(0)
                            covid = True
                            break
                    if 0 <= a - 1 < 5 and 0 <= b + 1 < 5 and room[a-1][b+1] == 'P':
                        if room[a - 1][b] == 'O' or room[a][b+1] == 'O':
                            answer.append(0)
                            covid = True
                            break
                    if 0 <= a + 1 < 5 and 0 <= b - 1 < 5 and room[a+1][b-1] == 'P':
                        if room[a+1][b] == 'O' or room[a][b-1] == 'O':
                            answer.append(0)
                            covid = True
                            break
                    if 0 <= a + 1 < 5 and 0 <= b + 1 < 5 and room[a+1][b+1] == 'P':
                        if room[a+1][b] == 'O' or room[a][b+1] == 'O':
                            answer.append(0)
                            covid = True
                            break
        if covid == False:
            answer.append(1)
        
        # a, b 가 있어 맨해튼 거리 2 이하 12 가지 경우
        # a, b - 1        a - 1, b    # 무조건 안됨
        # a, b + 1        a + 1, b    # 무조건 안됨
        # a, b - 2                    # a, b - 1이 x일 때 가능  
        # a - 2, b                    # a - 1, b 가 x 일 때 가능 
        # a, b + 2                    # a, b + 1이 x일 때 가능  
        # a + 2, b                    # a + 1, b 가 x 일 때 가능
        # a - 1, b - 1                # a - 1, b 가 x, a, b - 1 이 x 일 때 가능
        # a - 1, b + 1                # a - 1, b 가 x, a, b + 1 이 x 일 때 가능
        # a + 1, b - 1                # a + 1, b 가 x, a, b - 1 이 x 일 때 가능
        # a + 1, b + 1                # a + 1, b 가 x, a, b + 1 이 x 일 때 가능
        
    
    return answer

File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.java
import java.util.*;

class Solution {
    private static int[] dx = {-1, 1, 0, 0};
    private static int[] dy = {0, 0, -1, 1};
    
    public int[] solution(String[][] places) {
        int[] answer = new int[5];
        for (int r = 0; r < 5; r++) {
            char[][] room = new char[5][5];
            for (int i = 0; i < 5; i++) {
                room[i] = places[r][i].toCharArray();
            }
            if (isValidRoom(room)) answer[r] = 1;
            else answer[r] = 0;
        }
        return answer;
    }
    
    private boolean isValidRoom(char[][] room) {
        for (int i = 0; i < 5; i++) {
            for (int j = 0; j < 5; j++) {
                if (room[i][j] == 'P') {
                    if (!bfs(room, i, j)) return false;
                }
            }
        }
        return true;
    }
    
    private boolean bfs(char[][] room, int x, int y) {
        Queue<int[]> q = new LinkedList<>();
        boolean[][] visited = new boolean[5][5];
        q.add(new int[]{x, y, 0});
        visited[x][y] = true;
        
        while (!q.isEmpty()) {
            int[] current = q.poll();
            int cx = current[0], cy = current[1], dist = current[2];
            
            if (dist >= 2) continue;
            for (int i = 0; i < 4; i++) {
                int nx = cx + dx[i];
                int ny = cy + dy[i];
                if (nx < 0 || ny < 0 || nx >= 5 || ny >= 5 || visited[nx][ny] || room[nx][ny] == 'X') continue;
                if (room[nx][ny] == 'P') return false;
                q.add(new int[]{nx, ny, dist+1});
                visited[nx][ny] = true;
            }
                
        }
        return true;
    }
}